tests/Moq.Tests/InvocationsFixture.cs

```diff
@@ -45,11 +47,109 @@ public void MockInvocationsIncludeArguments()
 
 			var invocation = mock.Invocations[0];
 
-			var expectedArguments = new[] {obj};
+			var expectedArguments = new[] { obj };
 
 			Assert.Equal(expectedArguments, invocation.Arguments);
 		}
 
+		[Fact]
+		public void MockInvocationsIncludeReturnValue_NoSetup()
+		{
+			var mock = new Mock<IComparable>();
+
+			var obj = new object();
+
+			mock.Object.CompareTo(obj);
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(0, invocation.ReturnValue);
+			Assert.Null(invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeReturnValue_Setup()
+		{
+			var mock = new Mock<IComparable>();
+			var obj = new object();
+			mock.Setup(c => c.CompareTo(obj)).Returns(42);
+
+			mock.Object.CompareTo(obj);
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(42, invocation.ReturnValue);
+			Assert.Null(invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeReturnValue_BaseCall()
+		{
+			var mock = new Mock<Random>(1) // seed: 1
+			{
+				CallBase = true,
+			};
+
+			mock.Object.Next();
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(534011718, invocation.ReturnValue);
+			Assert.Null(invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeReturnValue_ReturnsException()
+		{
+			var mock = new Mock<ICloneable>();
+			var returnValue = new Exception();
+			mock.Setup(c => c.Clone()).Returns(returnValue);
+
+			mock.Object.Clone();
+
+			var invocation = mock.Invocations[0];
+
+			Assert.Equal(returnValue, invocation.ReturnValue);
+			Assert.Null(invocation.Exception);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeException_Setup()
+		{
+			var mock = new Mock<IComparable>();
+			var exception = new Exception("Message");
+			var exceptionThrownAndHandled = false;
+			mock.Setup(c => c.CompareTo(It.IsAny<object>())).Throws(exception);
+
+			try
+			{
+				mock.Object.CompareTo(null);
+			}
+			catch (Exception thrown)
+			{
+				Assert.Equal(exception.Message, thrown.Message);
+				var invocation = mock.Invocations[0];
+
+				Assert.Same(thrown, invocation.Exception);
+				exceptionThrownAndHandled = true;
+			}
+
+			Assert.True(exceptionThrownAndHandled);
+		}
+
+		[Fact]
+		public void MockInvocationsIncludeException_BaseCall()
+		{
+			var mock = new Mock<List<int>>()
+			{
+				CallBase = true,
+			};
+
+			Assert.Throws<ArgumentException>(() => mock.Object.GetRange(1, 1));
+			// Base-call exceptions are not recorded
+			Assert.Empty(mock.Invocations);
```

Comment:
Good of you for thinking of this test case! (I didn't.)

Something's odd here&mdash;invocations should get recorded even when `mock.CallBase == true`. I think the reason why no invocations get recorded here is that `List<>.GetRange` is not virtual.

Could you change this test so it calls some overridable method?

Reply:
Didn't think of the virtual requirement. I will add another test and change the code comment accordingly.

Reply:
I added another test 09816d01